datastore/statedata.go

```diff
@@ -227,9 +227,19 @@ type Artifact struct {
 		URI    string
 		Expire string
 	}
+	// Compatible devices for dependency checking.
 	CompatibleDevices []string `json:"device_types_compatible"`
-	ArtifactName      string   `json:"artifact_name"`
-	PayloadTypes      []string
+	// What kind of payloads are embedded in the artifact
+	// (e.g. rootfs-image).
+	PayloadTypes []string
+	// The following two properties implements ArtifactProvides header-info
+	// field of artifact version >= 3. The Attributes are moved to the root
+	// of the Artifact structure for backwards compatability.
+	ArtifactName  string `json:"artifact_name"`
+	ArtifactGroup string `json:"artifact_group"`
+	// This property is only used for deferring parsing of artifacts of
+	// version v3 and later. Is only read as nil value.
+	ArtifactProvidesBuffer json.RawMessage `json:"artifact_provides,omitempty"`
```

Comment:
I suspect it may be better to use `map[string]interface{}` here, so that you can index by string. This is also what `MergeArtifactProvides` returns.

Reply:
I think I'll keep the `header-info: artifact_depends` fields in the base structure for backwards compatibility, and dub this field to something like `TypeProvides map[string]interface{}` and pop the `artifact_name` and `artifact_group` upon parsing. This way we also avoid having to store duplicate fields in the DB. What do you think?

Reply:
Yes, that's a good idea. Just make sure to comment it clearly!